# Project 6: Analyzing Stock Sentiment from Twits
## Instructions
Each problem consists of a function to implement and instructions on how to implement the function.  The parts of the function that need to be implemented are marked with a `# TODO` comment.

## Packages
When you implement the functions, you'll only need to you use the packages you've used in the classroom, like [Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/). These packages will be imported for you. We recommend you don't add any import statements, otherwise the grader might not be able to run your code.

### Load Packages

In [1]:
import json
import nltk
import os
import random
import re
import torch

from torch import nn, optim
import torch.nn.functional as F

## Introduction
When deciding the value of a company, it's important to follow the news. For example, a product recall or natural disaster in a company's product chain. You want to be able to turn this information into a signal. Currently, the best tool for the job is a Neural Network. 

For this project, you'll use posts from the social media site [StockTwits](https://en.wikipedia.org/wiki/StockTwits). The community on StockTwits is full of investors, traders, and entrepreneurs. Each message posted is called a Twit. This is similar to Twitter's version of a post, called a Tweet. You'll build a model around these twits that generate a sentiment score.

We've collected a bunch of twits, then hand labeled the sentiment of each. To capture the degree of sentiment, we'll use a five-point scale: very negative, negative, neutral, positive, very positive. Each twit is labeled -2 to 2 in steps of 1, from very negative to very positive respectively. You'll build a sentiment analysis model that will learn to assign sentiment to twits on its own, using this labeled data.

The first thing we should to do, is load the data.

## Import Twits 
### Load Twits Data 
This JSON file contains a list of objects for each twit in the `'data'` field:

```
{'data':
  {'message_body': 'Neutral twit body text here',
   'sentiment': 0},
  {'message_body': 'Happy twit body text here',
   'sentiment': 1},
   ...
}
```

The fields represent the following:

* `'message_body'`: The text of the twit.
* `'sentiment'`: Sentiment score for the twit, ranges from -2 to 2 in steps of 1, with 0 being neutral.


To see what the data look like by printing the first 10 twits from the list. 

In [2]:
with open(os.path.join('..', '..', 'data', 'project_6_stocktwits', 'twits.json'), 'r') as f:
    twits = json.load(f)

print(twits['data'][:10])

[{'message_body': '$FITB great buy at 26.00...ill wait', 'sentiment': 2, 'timestamp': '2018-07-01T00:00:09Z'}, {'message_body': '@StockTwits $MSFT', 'sentiment': 1, 'timestamp': '2018-07-01T00:00:42Z'}, {'message_body': '#STAAnalystAlert for $TDG : Jefferies Maintains with a rating of Hold setting target price at USD 350.00. Our own verdict is Buy  http://www.stocktargetadvisor.com/toprating', 'sentiment': 2, 'timestamp': '2018-07-01T00:01:24Z'}, {'message_body': '$AMD I heard there’s a guy who knows someone who thinks somebody knows something - on StockTwits.', 'sentiment': 1, 'timestamp': '2018-07-01T00:01:47Z'}, {'message_body': '$AMD reveal yourself!', 'sentiment': 0, 'timestamp': '2018-07-01T00:02:13Z'}, {'message_body': '$AAPL Why the drop? I warren Buffet taking out his position?', 'sentiment': 1, 'timestamp': '2018-07-01T00:03:10Z'}, {'message_body': '$BA bears have 1 reason on 06-29 to pay more attention https://dividendbot.com?s=BA', 'sentiment': -2, 'timestamp': '2018-07-01T

### Length of Data
Now let's look at the number of twits in dataset. Print the number of twits below.

In [3]:
"""print out the number of twits"""

# TODO Implement 
print(len(twits['data']))


1548010


### Split Message Body and Sentiment Score

In [4]:
messages = [twit['message_body'] for twit in twits['data']]
# Since the sentiment scores are discrete, we'll scale the sentiments to 0 to 4 for use in our network
sentiments = [twit['sentiment'] + 2 for twit in twits['data']]

## Preprocessing the Data
With our data in hand we need to preprocess our text. These twits are collected by filtering on ticker symbols where these are denoted with a leader $ symbol in the twit itself. For example,

`{'message_body': 'RT @google Our annual look at the year in Google blogging (and beyond) http://t.co/sptHOAh8 $GOOG',
 'sentiment': 0}`

The ticker symbols don't provide information on the sentiment, and they are in every twit, so we should remove them. This twit also has the `@google` username, again not providing sentiment information, so we should also remove it. We also see a URL `http://t.co/sptHOAh8`. Let's remove these too.

The easiest way to remove specific words or phrases is with regex using the `re` module. You can sub out specific patterns with a space:

```python
re.sub(pattern, ' ', text)
```
This will substitute a space with anywhere the pattern matches in the text. Later when we tokenize the text, we'll split appropriately on those spaces.

In [5]:
#for i in range(10):
#    text = messages[i]
#    print('msg', i, ':', text)
#    text = text.lower()
#    text = re.sub(r'https?://[^\s]+', ' ', text)
#    text = re.sub(r'[$][A-Za-z][\S]*', ' ', text)
#    text = re.sub(r'[@][A-Za-z][\S]*', ' ', text)
#    text = re.sub('[^a-zA-Z]+', ' ', text)
#    print('msg', i, ':', text)
#    print('msg', i, ':', text.split())
#    print('------')

### Pre-Processing

In [6]:
nltk.download('wordnet')


def preprocess(message):
    """
    This function takes a string as input, then performs these operations: 
        - lowercase
        - remove URLs
        - remove ticker symbols 
        - removes punctuation
        - tokenize by splitting the string on whitespace 
        - removes any single character tokens
    
    Parameters
    ----------
        message : The text message to be preprocessed.
        
    Returns
    -------
        tokens: The preprocessed text into tokens.
    """ 
    #TODO: Implement 
    
    # Lowercase the twit message
    text = message.lower()
    
    # Replace URLs with a space in the message
    text = re.sub(r'https?://[^\s]+', ' ', text)
    
    # Replace ticker symbols with a space. The ticker symbols are any stock symbol that starts with $.
    text = re.sub(r'[$][A-Za-z][\S]*', ' ', text)
    
    # Replace StockTwits usernames with a space. The usernames are any word that starts with @.
    text = re.sub(r'[@][A-Za-z][\S]*', ' ', text)

    # Replace everything not a letter with a space
    text = re.sub('[^a-zA-Z]+', ' ', text)
    
    # Tokenize by splitting the string on whitespace into a list of words
    tokens = text.split()

    # Lemmatize words using the WordNetLemmatizer. You can ignore any word that is not longer than one character.
    wnl = nltk.stem.WordNetLemmatizer()
    tokens = [wnl.lemmatize(word) for word in tokens]

    
    assert type(tokens) == list, 'Tokens should be list'
    return tokens

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


>Note: You must ensure that after preprocessing the text should NOT include:
- Numbers
- URLs
- Single character tokens
- Ticker symbols (these should be removed even if they don't appear at the beginning)

### Preprocess All the Twits 
Now we can preprocess each of the twits in our dataset. Apply the function `preprocess` to all the twit messages.

In [7]:
# TODO Implement

tokenized = [preprocess(line) for line in messages]

In [8]:
for i in range(4):
    print(messages[i])
    print(tokenized[i])
    print('-------')
print(len(tokenized))

$FITB great buy at 26.00...ill wait
['great', 'buy', 'at', 'ill', 'wait']
-------
@StockTwits $MSFT
[]
-------
#STAAnalystAlert for $TDG : Jefferies Maintains with a rating of Hold setting target price at USD 350.00. Our own verdict is Buy  http://www.stocktargetadvisor.com/toprating
['staanalystalert', 'for', 'jefferies', 'maintains', 'with', 'a', 'rating', 'of', 'hold', 'setting', 'target', 'price', 'at', 'usd', 'our', 'own', 'verdict', 'is', 'buy']
-------
$AMD I heard there’s a guy who knows someone who thinks somebody knows something - on StockTwits.
['i', 'heard', 'there', 's', 'a', 'guy', 'who', 'know', 'someone', 'who', 'think', 'somebody', 'know', 'something', 'on', 'stocktwits']
-------
1548010


### Bag of Words
Now with all of our messages tokenized, we want to create a vocabulary and count up how often each word appears in our entire corpus. Use the [`Counter`](https://docs.python.org/3.1/library/collections.html#collections.Counter) function to count up all the tokens.

In [9]:
from collections import Counter


"""
Create a vocabulary by using Bag of words
"""

# TODO: Implement 
vocabularies = []
for words in tokenized:
    for word in words:
        vocabularies.append(word)

bow = Counter(vocabularies)

In [10]:
list(bow.items())[:5]

[('great', 17568),
 ('buy', 76321),
 ('at', 138455),
 ('ill', 882),
 ('wait', 11370)]

### Frequency of Words Appearing in Message
With our vocabulary, now we'll remove some of the most common words such as 'the', 'and', 'it', etc. These words don't contribute to identifying sentiment and are really common, resulting in a lot of noise in our input. If we can filter these out, then our network should have an easier time learning.

We also want to remove really rare words that show up in a only a few twits. Here you'll want to divide the count of each word by the **number of messages** calculated in the code block above (i.e. `len(messages))`. Then remove words that only appear in some small fraction of the messages.

>Note: There is not an exact number for low and high-frequency cut-offs, however there is a correct optimal range.
You should ideally set up low-frequency cut-off from 0.0000002 to 0.000007 (inclusive) and high-frequency from 5 to 20 (inclusive). If the number is too big, we lose lots of important words that we can use in our data.

In [11]:
"""
Set the following variables:
    freqs
    low_cutoff
    high_cutoff
    K_most_common
"""

# TODO Implement 

total_count = len(messages)

# Dictionart that contains the Frequency of words appearing in messages.
# The key is the token and the value is the frequency of that word in the corpus.
freqs = {word: count/total_count for word, count in bow.items()}

# Float that is the frequency cutoff. Drop words with a frequency that is lower or equal to this number.
low_cutoff = 0.000007

# Integer that is the cut off for most common words. Drop words that are the `high_cutoff` most common words.
high_cutoff = 20

# The k most common words in the corpus. Use `high_cutoff` as the k.
K_most_common = bow.most_common(high_cutoff)

filtered_words = [word for word in freqs if (freqs[word] > low_cutoff and word not in K_most_common)]
print(K_most_common)
len(filtered_words) 

[('the', 398778), ('to', 379505), ('a', 325404), ('is', 284879), ('for', 273543), ('on', 241665), ('of', 211334), ('and', 208492), ('in', 205321), ('this', 203562), ('s', 193532), ('it', 193504), ('i', 182940), ('at', 138455), ('will', 128199), ('up', 121581), ('are', 101426), ('you', 94290), ('that', 89668), ('be', 89278)]


18153

In [12]:
print(list(freqs.items())[:10])
print(len(freqs))
print(filtered_words[:10])
print(len(filtered_words))

[('great', 0.011348763896874052), ('buy', 0.04930265308363641), ('at', 0.08944063668839349), ('ill', 0.0005697637612160128), ('wait', 0.007344913792546559), ('staanalystalert', 0.00263047396334649), ('for', 0.1767062228280179), ('jefferies', 0.0006821661358776752), ('maintains', 0.0005833295650544894), ('with', 0.04448227078636443)]
98474
['great', 'buy', 'at', 'ill', 'wait', 'staanalystalert', 'for', 'jefferies', 'maintains', 'with']
18153


### Updating Vocabulary by Removing Filtered Words
Let's creat three variables that will help with our vocabulary.

In [13]:
"""
Set the following variables:
    vocab
    id2vocab
    filtered
"""

#TODO Implement

# A dictionary for the `filtered_words`. The key is the word and value is an id that represents the word. 
vocab = {filtered_words[i]: i for i in range(len(filtered_words))}
# Reverse of the `vocab` dictionary. The key is word id and value is the word. 
id2vocab = {i: filtered_words[i] for i in range(len(filtered_words))}
# tokenized with the words not in `filtered_words` removed.
filtered = []
n = 0
for words in tokenized:
    filtered.append([word for word in words if word in filtered_words])

assert set(vocab.keys()) == set(id2vocab.values()), 'Check vocab and id2vocab dictionaries'

In [14]:
filtered[:3]

[['great', 'buy', 'at', 'ill', 'wait'],
 [],
 ['staanalystalert',
  'for',
  'jefferies',
  'maintains',
  'with',
  'a',
  'rating',
  'of',
  'hold',
  'setting',
  'target',
  'price',
  'at',
  'usd',
  'our',
  'own',
  'verdict',
  'is',
  'buy']]

### Balancing the classes
Let's do a few last pre-processing steps. If we look at how our twits are labeled, we'll find that 50% of them are neutral. This means that our network will be 50% accurate just by guessing 0 every single time. To help our network learn appropriately, we'll want to balance our classes.
That is, make sure each of our different sentiment scores show up roughly as frequently in the data.

What we can do here is go through each of our examples and randomly drop twits with neutral sentiment. What should be the probability we drop these twits if we want to get around 20% neutral twits starting at 50% neutral? We should also take this opportunity to remove messages with length 0.

In [15]:
balanced = {'messages': [], 'sentiments':[]}

n_neutral = sum(1 for each in sentiments if each == 2)
N_examples = len(sentiments)
keep_prob = (N_examples - n_neutral)/4/n_neutral

for idx, sentiment in enumerate(sentiments):
    message = filtered[idx]
    if len(message) == 0:
        # skip this message because it has length zero
        continue
    elif sentiment != 2 or random.random() < keep_prob:
        balanced['messages'].append(message)
        balanced['sentiments'].append(sentiment) 

If you did it correctly, you should see the following result 

In [16]:
n_neutral = sum(1 for each in balanced['sentiments'] if each == 2)
N_examples = len(balanced['sentiments'])
n_neutral/N_examples

0.19450558725064973

Finally let's convert our tokens into integer ids which we can pass to the network.

In [17]:
token_ids = [[vocab[word] for word in message] for message in balanced['messages']]
sentiments = balanced['sentiments']

## Neural Network
Now we have our vocabulary which means we can transform our tokens into ids, which are then passed to our network. So, let's define the network now!

Here is a nice diagram showing the network we'd like to build: 

#### Embed -> RNN -> Dense -> Softmax
### Implement the text classifier
Before we build text classifier, if you remember from the other network that you built in  "Sentiment Analysis with an RNN"  exercise  - which there, the network called " SentimentRNN", here we named it "TextClassifer" - consists of three main parts: 1) init function `__init__` 2) forward pass `forward`  3) hidden state `init_hidden`. 

This network is pretty similar to the network you built expect in the  `forward` pass, we use softmax instead of sigmoid. The reason we are not using sigmoid is that the output of NN is not a binary. In our network, sentiment scores have 5 possible outcomes. We are looking for an outcome with the highest probability thus softmax is a better choice.

In [18]:
# First checking if GPU is available
train_on_gpu=torch.cuda.is_available()

if(train_on_gpu):
    print('Training on GPU.')
else:
    print('No GPU available, training on CPU.')

Training on GPU.


In [19]:
class TextClassifier(nn.Module):
    def __init__(self, vocab_size, embed_size, lstm_size, output_size, lstm_layers=1, dropout=0.1):
        """
        Initialize the model by setting up the layers.
        
        Parameters
        ----------
            vocab_size : The vocabulary size.
            embed_size : The embedding layer size.
            lstm_size : The LSTM layer size.
            output_size : The output size.
            lstm_layers : The number of LSTM layers.
            dropout : The dropout probability.
        """
        
        super().__init__()
        self.vocab_size = vocab_size
        self.embed_size = embed_size
        self.lstm_size = lstm_size
        self.output_size = output_size
        self.lstm_layers = lstm_layers
        self.dropout = dropout
        
        # TODO Implement

        # Setup embedding layer
        self.embedding = nn.Embedding(vocab_size, embed_size)
        
        # Setup additional layers
        self.lstm = nn.LSTM(embed_size, lstm_size, lstm_layers, dropout=dropout, batch_first=False)
        self.dropout = nn.Dropout(dropout)
        self.fc = nn.Linear(lstm_size, output_size)
        self.soft = nn.LogSoftmax(dim=1)


    def init_hidden(self, batch_size):
        """ 
        Initializes hidden state
        
        Parameters
        ----------
            batch_size : The size of batches.
        
        Returns
        -------
            hidden_state
            
        """
        
        # TODO Implement 
        
        # Create two new tensors with sizes n_layers x batch_size x hidden_dim,
        # initialized to zero, for hidden state and cell state of LSTM
        
        weight = next(self.parameters()).data

        if (train_on_gpu):
            hidden = (weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_().cuda(),
                      weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_().cuda())
        else:
            hidden = (weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_(),
                      weight.new(self.lstm_layers, batch_size, self.lstm_size).zero_())

        return hidden


    def forward(self, nn_input, hidden_state):
        """
        Perform a forward pass of our model on nn_input.
        
        Parameters
        ----------
            nn_input : The batch of input to the NN.
            hidden_state : The LSTM hidden state.

        Returns
        -------
            logps: log softmax output
            hidden_state: The new hidden state.

        """
        
        # TODO Implement 
        batch_size = nn_input.size(0)

        # embeddings and lstm_out
        embeds = self.embedding(nn_input)
        lstm_out, hidden_state = self.lstm(embeds, hidden_state)

        # stack up lstm outputs
        lstm_out = lstm_out.contiguous().view(-1, self.lstm_size)

        # dropout and fully-connected layer
        out = self.fc(lstm_out)
        out = self.dropout(out)

        # sigmoid function
        sig_out = self.soft(out)

        # reshape to be batch_size first
        sig_out = sig_out.view(batch_size, -1, self.output_size)
        sig_out = sig_out[-1] # get last batch of labels

        # return last sigmoid output and hidden state
        return sig_out, hidden_state

### View Model

In [20]:
model = TextClassifier(len(vocab), 10, 6, 5, dropout=0.1, lstm_layers=2)
model.embedding.weight.data.uniform_(-1, 1)
input = torch.randint(0, 1000, (5, 4), dtype=torch.int64)
hidden = model.init_hidden(4)

if train_on_gpu:
    model = model.cuda()
    input = input.cuda()

logps, _ = model.forward(input, hidden)
print(logps)

tensor([[-1.8328, -1.5276, -1.6559, -1.4224, -1.6559],
        [-1.8115, -1.5203, -1.6884, -1.4167, -1.6575],
        [-1.7617, -1.7915, -1.5743, -1.3881, -1.5856],
        [-1.8086, -1.5069, -1.6744, -1.4451, -1.6536]], device='cuda:0')


## Training
### DataLoaders and Batching
Now we should build a generator that we can use to loop through our data. It'll be more efficient if we can pass our sequences in as batches. Our input tensors should look like `(sequence_length, batch_size)`. So if our sequences are 40 tokens long and we pass in 25 sequences, then we'd have an input size of `(40, 25)`.

If we set our sequence length to 40, what do we do with messages that are more or less than 40 tokens? For messages with fewer than 40 tokens, we will pad the empty spots with zeros. We should be sure to **left** pad so that the RNN starts from nothing before going through the data. If the message has 20 tokens, then the first 20 spots of our 40 long sequence will be 0. If a message has more than 40 tokens, we'll just keep the first 40 tokens.

In [21]:
def dataloader(messages, labels, sequence_length=30, batch_size=32, shuffle=False):
    """ 
    Build a dataloader.
    """
    if shuffle:
        indices = list(range(len(messages)))
        random.shuffle(indices)
        messages = [messages[idx] for idx in indices]
        labels = [labels[idx] for idx in indices]

    total_sequences = len(messages)

    for ii in range(0, total_sequences, batch_size):
        batch_messages = messages[ii: ii+batch_size]
        
        # First initialize a tensor of all zeros
        batch = torch.zeros((sequence_length, len(batch_messages)), dtype=torch.int64)
        for batch_num, tokens in enumerate(batch_messages):
            token_tensor = torch.tensor(tokens)
            # Left pad!
            start_idx = max(sequence_length - len(token_tensor), 0)
            batch[start_idx:, batch_num] = token_tensor[:sequence_length]
        
        label_tensor = torch.tensor(labels[ii: ii+len(batch_messages)])
        
        yield batch, label_tensor

### Training and  Validation
With our data in nice shape, we'll split it into training and validation sets.

In [22]:
len(sentiments)

1030829

In [23]:
"""
Split data into training and validation datasets. Use an appropriate split size.
The features are the `token_ids` and the labels are the `sentiments`.
"""   

# TODO Implement 
split_size = round(0.7 * len(token_ids))

train_features = token_ids[:split_size]
valid_features = token_ids[split_size:]
train_labels = sentiments[:split_size]
valid_labels = sentiments[split_size:]

In [24]:
text_batch, labels = next(iter(dataloader(train_features, train_labels, sequence_length=20, batch_size=64)))
model = TextClassifier(len(vocab)+1, 200, 128, 5, dropout=0.)

if train_on_gpu:
    model = model.cuda()
    text_batch = text_batch.cuda()

hidden = model.init_hidden(64)
logps, hidden = model.forward(text_batch, hidden)

### Training
It's time to train the neural network!

In [25]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = TextClassifier(len(vocab)+1, 1024, 512, 5, lstm_layers=2, dropout=0.2)
model.embedding.weight.data.uniform_(-1, 1)
model.to(device)

TextClassifier(
  (embedding): Embedding(18154, 1024)
  (lstm): LSTM(1024, 512, num_layers=2, dropout=0.2)
  (dropout): Dropout(p=0.2)
  (fc): Linear(in_features=512, out_features=5, bias=True)
  (soft): LogSoftmax()
)

In [26]:
"""
Train your model with dropout. Make sure to clip your gradients.
Print the training loss, validation loss, and validation accuracy for every 100 steps.
"""

epochs = 4
batch_size = 256
learning_rate = 0.001

print_every = 100
#print_every = 1
criterion = nn.NLLLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
model.train()

for epoch in range(epochs):
    print('Starting epoch {}'.format(epoch + 1))
    
    steps = 0
    
    # =========FROM EXERCISE=========
    # initialize hidden state
    hidden = model.init_hidden(batch_size)
    # =========FROM EXERCISE=========
    
    for text_batch, labels in dataloader(
            train_features, train_labels, batch_size=batch_size, sequence_length=20):
        steps += 1
        hidden = tuple([each.data for each in hidden])
                
        # Set Device
        text_batch, labels = text_batch.to(device), labels.to(device)
        if text_batch.size() != torch.Size([20, batch_size]):
            continue
        for each in hidden:
            each.to(device)
        
        # TODO Implement: Train Model
        
        # zero accumulated gradients
        model.zero_grad()

        # get the output from the model
        output, hidden = model.forward(text_batch, hidden)

        # calculate the loss and perform backprop
        loss = criterion(output, labels)
        loss.backward()
        # `clip_grad_norm` helps prevent the exploding gradient problem in RNNs / LSTMs.
        nn.utils.clip_grad_norm_(model.parameters(), 5)
        optimizer.step()
        
        if steps % print_every == 0:
            model.eval()
            
            # TODO Implement: Print metrics
            
            # Get validation loss
            val_h = model.init_hidden(batch_size)
            val_losses = []
            val_accuracy = []
            accuracy = 0
            
            with torch.no_grad():
                for val_batch, val_labels in dataloader(valid_features, 
                        valid_labels, batch_size=batch_size, sequence_length=20):

                    # Creating new variables for the hidden state, otherwise
                    # we'd backprop through the entire training history
                    
                    #val_h = tuple([each.data for each in val_h])

                    #if(train_on_gpu):
                    #    inputs, labels = inputs.cuda(), labels.cuda()
                    val_batch, val_labels = val_batch.to(device), val_labels.to(device)
                    if val_batch.size() != torch.Size([20, batch_size]):
                        continue
                    for each in val_h:
                        each.to(device)

                    val_output, val_h = model.forward(val_batch, val_h)
                    val_loss = criterion(val_output, val_labels)

                    val_losses.append(val_loss.item())
                    
                    ps = torch.exp(val_output)
                    top_p, top_class = ps.topk(1, dim=1)
                    equals = top_class == val_labels.view(*top_class.shape)
                    accuracy += torch.mean(equals.type(torch.FloatTensor))

                model.train()
                print("Epoch: {}/{}...".format(epoch+1, epochs),
                      "Step: {}...".format(steps),
                      "Loss: {:.6f}...".format(loss.item()),
                      "Val Loss: {:.6f}".format(sum(val_losses)/len(val_losses)),
                      "Val Accuracy: {:.3f}".format(accuracy/len(val_losses)))

Starting epoch 1
Epoch: 1/4... Step: 100... Loss: 1.178489... Val Loss: 1.179000 Val Accuracy: 0.517
Epoch: 1/4... Step: 200... Loss: 1.125943... Val Loss: 1.085643 Val Accuracy: 0.553
Epoch: 1/4... Step: 300... Loss: 0.872893... Val Loss: 1.010657 Val Accuracy: 0.607
Epoch: 1/4... Step: 400... Loss: 0.725083... Val Loss: 0.964654 Val Accuracy: 0.626
Epoch: 1/4... Step: 500... Loss: 1.124666... Val Loss: 0.919156 Val Accuracy: 0.640
Epoch: 1/4... Step: 600... Loss: 0.936631... Val Loss: 0.905477 Val Accuracy: 0.650
Epoch: 1/4... Step: 700... Loss: 1.080026... Val Loss: 0.887894 Val Accuracy: 0.661
Epoch: 1/4... Step: 800... Loss: 0.851160... Val Loss: 0.871587 Val Accuracy: 0.659
Epoch: 1/4... Step: 900... Loss: 1.057556... Val Loss: 0.850673 Val Accuracy: 0.676
Epoch: 1/4... Step: 1000... Loss: 0.932696... Val Loss: 0.840885 Val Accuracy: 0.677
Epoch: 1/4... Step: 1100... Loss: 0.741879... Val Loss: 0.823480 Val Accuracy: 0.683
Epoch: 1/4... Step: 1200... Loss: 0.714557... Val Loss: 0

Epoch: 4/4... Step: 1400... Loss: 0.648604... Val Loss: 0.768771 Val Accuracy: 0.714
Epoch: 4/4... Step: 1500... Loss: 0.704927... Val Loss: 0.765425 Val Accuracy: 0.714
Epoch: 4/4... Step: 1600... Loss: 0.691407... Val Loss: 0.793196 Val Accuracy: 0.711
Epoch: 4/4... Step: 1700... Loss: 0.775665... Val Loss: 0.772042 Val Accuracy: 0.714
Epoch: 4/4... Step: 1800... Loss: 0.689932... Val Loss: 0.797547 Val Accuracy: 0.710
Epoch: 4/4... Step: 1900... Loss: 0.763173... Val Loss: 0.778069 Val Accuracy: 0.712
Epoch: 4/4... Step: 2000... Loss: 0.323299... Val Loss: 0.779612 Val Accuracy: 0.715
Epoch: 4/4... Step: 2100... Loss: 0.701683... Val Loss: 0.782944 Val Accuracy: 0.714
Epoch: 4/4... Step: 2200... Loss: 0.697966... Val Loss: 0.784590 Val Accuracy: 0.713
Epoch: 4/4... Step: 2300... Loss: 0.568517... Val Loss: 0.789596 Val Accuracy: 0.710
Epoch: 4/4... Step: 2400... Loss: 0.644830... Val Loss: 0.783356 Val Accuracy: 0.709
Epoch: 4/4... Step: 2500... Loss: 0.642808... Val Loss: 0.782533 

In [27]:
torch.save(model.state_dict(), 'checkpoint.pth')

## Making Predictions
### Prediction 
Okay, now that you have a trained model, try it on some new twits and see if it works appropriately. Remember that for any new text, you'll need to preprocess it first before passing it to the network. Implement the `predict` function to generate the prediction vector from a message.

In [43]:
def predict(text, model, vocab):
    """ 
    Make a prediction on a single sentence.

    Parameters
    ----------
        text : The string to make a prediction on.
        model : The model to use for making the prediction.
        vocab : Dictionary for word to word ids. The key is the word and the value is the word id.

    Returns
    -------
        pred : Prediction vector
    """    
    
    # TODO Implement
    
    tokens = preprocess(text)
    
    # Filter non-vocab words
    #print('-----1')
    tokens = [word for word in tokens if word in filtered_words]
    #print('-----2')
    # Convert words to ids
    tokens =  [vocab[word] for word in tokens]
    #print('-----3')
        
    # Adding a batch dimension
    #print(torch.tensor(tokens))
    #print(torch.tensor(tokens).unsqueeze(1))
    text_input = torch.tensor(tokens).unsqueeze(1)
    # Get the NN output
    hidden = model.init_hidden(1)
    #print(len(text_input))
    #print(len(hidden))
    #print(hidden)
    logps, hidden = model.forward(text_input, hidden)
    # Take the exponent of the NN output to get a range of 0 to 1 for each label.
    pred = torch.exp(logps)
    
    return pred

In [44]:
text = "Google is working on self driving cars, I'm bullish on $goog"
model.eval()
model.to("cpu")
train_on_gpu = False
predict(text, model, vocab)

tensor([[ 0.0003,  0.0134,  0.0087,  0.8882,  0.0894]])

### Questions: What is the prediction of the model? What is the uncertainty of the prediction?
** TODO: Answer Question**

The output shows a 1x5 matrix, in which the components correspond to very negative, negative, neutral, positive, and very positive. The probable sentiment of this twit is "positive" with the rate at 88.82% and the second high is "very positive" (8.94%). The "negative" along with "very negative" is only 1.37%, which is less than 5% (2-sigma confidence level), showing that the model is very certain at its prediction. The prediction fits our feeling from the wordings.

Now we have a trained model and we can make predictions. We can use this model to track the sentiments of various stocks by predicting the sentiments of twits as they are coming in. Now we have a stream of twits. For each of those twits, pull out the stocks mentioned in them and keep track of the sentiments. Remember that in the twits, ticker symbols are encoded with a dollar sign as the first character, all caps, and 2-4 letters, like $AAPL. Ideally, you'd want to track the sentiments of the stocks in your universe and use this as a signal in your larger model(s).

## Testing
### Load the Data 

In [45]:
with open(os.path.join('..', '..', 'data', 'project_6_stocktwits', 'test_twits.json'), 'r') as f:
    test_data = json.load(f)

### Twit Stream

In [46]:
def twit_stream():
    for twit in test_data['data']:
        yield twit

next(twit_stream())

{'message_body': '$JWN has moved -1.69% on 10-31. Check out the movement and peers at  https://dividendbot.com?s=JWN',
 'timestamp': '2018-11-01T00:00:05Z'}

Using the `prediction` function, let's apply it to a stream of twits.

In [47]:
def score_twits(stream, model, vocab, universe):
    """ 
    Given a stream of twits and a universe of tickers, return sentiment scores for tickers in the universe.
    """
    for twit in stream:

        # Get the message text
        text = twit['message_body']
        symbols = re.findall('\$[A-Z]{2,4}', text)
        score = predict(text, model, vocab)

        for symbol in symbols:
            if symbol in universe:
                yield {'symbol': symbol, 'score': score, 'timestamp': twit['timestamp']}

In [48]:
universe = {'$BBRY', '$AAPL', '$AMZN', '$BABA', '$YHOO', '$LQMT', '$FB', '$GOOG', '$BBBY', '$JNUG', '$SBUX', '$MU'}
score_stream = score_twits(twit_stream(), model, vocab, universe)

next(score_stream)

{'symbol': '$AAPL',
 'score': tensor([[ 0.0732,  0.0955,  0.2575,  0.3149,  0.2589]]),
 'timestamp': '2018-11-01T00:00:18Z'}

That's it. You have successfully built a model for sentiment analysis! 

## Submission
Now that you're done with the project, it's time to submit it. Click the submit button in the bottom right. One of our reviewers will give you feedback on your project with a pass or not passed grade. You can continue to the next section while you wait for feedback.